xline/src/storage/authstore/backend.rs

```diff
@@ -0,0 +1,1235 @@
+        let kv = KeyValue {
+            key,
+            value,
+            create_revision: new_rev.create_revision,
+            mod_revision: new_rev.mod_revision,
+            version: new_rev.version,
+            ..KeyValue::default()
+        };
+        let _prev = self.db.insert(new_rev.as_revision(), kv);
+    }
+
+    /// put user to `AuthStore`
+    fn put_user(&self, user: &User, revision: i64, sub_revision: i64) {
+        let key = [USER_PREFIX, &user.name].concat();
+        let value = user.encode_to_vec();
+        self.put(key, value, revision, sub_revision);
+    }
+
+    /// put role to `AuthStore`
+    fn put_role(&self, role: &Role, revision: i64, sub_revision: i64) {
+        let key = [ROLE_PREFIX, &role.name].concat();
+        let value = role.encode_to_vec();
+        self.put(key, value, revision, sub_revision);
+    }
+
+    /// delete `KeyValue` in `AuthStore`
+    fn delete(&self, key: &[u8], revision: i64, sub_revision: i64) {
+        let revisions = self.index.delete(key, &[], revision, sub_revision);
+        let _prev_kv = self.db.mark_deletions(&revisions);
+    }
+
+    /// delete user in `AuthStore`
+    fn delete_user(&self, username: &str, revision: i64, sub_revision: i64) {
+        let key = [USER_PREFIX, username.as_bytes()].concat();
+        self.delete(&key, revision, sub_revision);
+    }
+
+    /// delete role in `AuthStore`
+    fn delete_role(&self, rolename: &str, revision: i64, sub_revision: i64) {
+        let key = [ROLE_PREFIX, rolename.as_bytes()].concat();
+        self.delete(&key, revision, sub_revision);
+    }
+
+    /// Get all users in the `AuthStore`
+    fn get_all_users(&self) -> Vec<User> {
+        let range_end = KeyRange::get_prefix(USER_PREFIX);
+        let revisions = self.index.get(USER_PREFIX, &range_end, 0);
+        self.db
+            .get_values(&revisions)
+            .into_iter()
+            .map(|kv| {
+                User::decode(kv.value.as_slice()).unwrap_or_else(|e| {
+                    panic!(
+                        "Failed to decode user from kv value, error: {:?}, kv: {:?}",
+                        e, kv
+                    )
+                })
+            })
+            .collect()
+    }
+
+    /// Get all roles in the `AuthStore`
+    fn get_all_roles(&self) -> Vec<Role> {
+        let range_end = KeyRange::get_prefix(ROLE_PREFIX);
+        let revisions = self.index.get(ROLE_PREFIX, &range_end, 0);
+        self.db
+            .get_values(&revisions)
+            .into_iter()
+            .map(|kv| {
+                Role::decode(kv.value.as_slice()).unwrap_or_else(|e| {
+                    panic!(
+                        "Failed to decode role from kv value, error: {:?}, kv: {:?}",
+                        e, kv
+                    )
+                })
+            })
+            .collect()
+    }
+
+    /// speculative execute command
+    pub(crate) fn speculative_exec(&self, execution_req: ExecutionRequest) {
```

Comment:
Why do we call it "speculative_exec"? How is it speculative?

Reply:
just use old code, command execution will be refactor in the future